provider/plugin/plugin.go

```diff
@@ -0,0 +1,350 @@
+		if resp.StatusCode >= 300 && resp.StatusCode < 500 {
+			return backoff.Permanent(fmt.Errorf("status code < 500"))
+		}
+		return nil
+	}, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), maxRetries))
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to plugin api: %v", err)
+	}
+
+	vary := resp.Header.Get(varyHeader)
+	contentType := resp.Header.Get(contentTypeHeader)
+
+	if vary != contentTypeHeader {
+		return nil, fmt.Errorf("wrong vary value returned from server: %s", vary)
+	}
+
+	if contentType != mediaTypeFormatAndVersion {
+		return nil, fmt.Errorf("wrong content type returned from server: %s", contentType)
+	}
+
+	return &PluginProvider{
+		client:          client,
+		remoteServerURL: parsedURL,
+	}, nil
+}
+
+// Records will make a GET call to remoteServerURL/records and return the results
+func (p PluginProvider) Records(ctx context.Context) ([]*endpoint.Endpoint, error) {
+	u, err := url.JoinPath(p.remoteServerURL.String(), "records")
+	if err != nil {
+		recordsErrorsGauge.Inc()
+		log.Debugf("Failed to join path: %s", err.Error())
+		return nil, err
+	}
+	req, err := http.NewRequest("GET", u, nil)
+	if err != nil {
+		recordsErrorsGauge.Inc()
+		log.Debugf("Failed to create request: %s", err.Error())
+		return nil, err
+	}
+	req.Header.Set(acceptHeader, mediaTypeFormatAndVersion)
+	resp, err := p.client.Do(req)
+	if err != nil {
+		recordsErrorsGauge.Inc()
+		log.Debugf("Failed to perform request: %s", err.Error())
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		recordsErrorsGauge.Inc()
+		log.Debugf("Failed to get records with code %d", resp.StatusCode)
+		return nil, fmt.Errorf("failed to get records with code %d", resp.StatusCode)
+	}
+
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		recordsErrorsGauge.Inc()
+		log.Debugf("Failed to read response body: %s", err.Error())
+		return nil, err
+	}
+
+	endpoints := []*endpoint.Endpoint{}
+	err = json.Unmarshal(b, &endpoints)
+	if err != nil {
+		recordsErrorsGauge.Inc()
+		log.Debugf("Failed to unmarshal response body: %s", err.Error())
+		return nil, err
+	}
+	return endpoints, nil
+}
+
+// ApplyChanges will make a POST to remoteServerURL/records with the changes
+func (p PluginProvider) ApplyChanges(ctx context.Context, changes *plan.Changes) error {
+	u, err := url.JoinPath(p.remoteServerURL.String(), "records")
+	if err != nil {
+		applyChangesErrorsGauge.Inc()
+		log.Debugf("Failed to join path: %s", err.Error())
+		return err
+	}
+	b, err := json.Marshal(changes)
+	if err != nil {
+		applyChangesErrorsGauge.Inc()
+		log.Debugf("Failed to marshal changes: %s", err.Error())
+		return err
+	}
+
+	req, err := http.NewRequest("POST", u, bytes.NewBuffer(b))
+	if err != nil {
+		applyChangesErrorsGauge.Inc()
+		log.Debugf("Failed to create request: %s", err.Error())
+		return err
+	}
+	req.Header.Set(acceptHeader, mediaTypeFormatAndVersion)
```

Comment:
Not necessary as we're not getting a response body back.

Reply:
Yes, as I said in a previous comment, this was only for integrating with the IONOS plugin which seems to have implemented the spec wrong. I plan to remove those. 

/cc @akrieg-ionos

Reply:
ok, as @ePaul has [clarified the details now](https://github.com/kubernetes-sigs/external-dns/pull/3063#discussion_r1219352975) , we know now which headers must be set. Regarding the server, we will create a new release of [ionos-plugin](https://github.com/ionos-cloud/external-dns-ionos-plugin) which fullfills these requirements.

Reply:
@Raffo @ePaul , as mentioned above we have now a [new ionos-plugin version](https://github.com/ionos-cloud/external-dns-ionos-plugin/releases/tag/v0.2.0).
In the [server_test.go](https://github.com/ionos-cloud/external-dns-ionos-plugin/blob/main/cmd/plugin/init/server/server_test.go) you can see the exact server behavior regarding version headers.

Reply:
Thanks, I will re-run the tests.